src/compass/utils/validate_cslc.py

```diff
@@ -110,12 +110,33 @@ def compare_cslc_products(file_ref, file_sec):
         return
 
     # Compare amplitude of reference and generated CSLC products
-    print('Check mean real part difference between CSLC products is < 1.0e-5')
-    assert np.allclose(slc_ref.real, slc_sec.real,
-                       atol=0.0, rtol=1.0e-5, equal_nan=True)
-    print('Check mean imaginary part difference between CSLC products is < 1.0e-5')
-    assert np.allclose(slc_ref.imag, slc_sec.imag,
-                       atol=0.0, rtol=1.0e-5, equal_nan=True)
+    np.seterr(invalid='ignore')
+    diff_real = np.abs((slc_ref.real - slc_sec.real) / slc_sec.real)
+    diff_imag =  np.abs((slc_ref.imag - slc_sec.imag) / slc_sec.imag)
+
+    # Compute total number of pixels different from nan from ref and sec
+    tot_pixels_ref = np.count_nonzero(~np.isnan(np.abs(slc_ref)))
+    tot_pixels_sec = np.count_nonzero(~np.isnan(np.abs(slc_ref)))
+
+    # Check that total number of pixel is the same
+    assert tot_pixels_ref == tot_pixels_sec
+
+    # Compute the number of pixels in real part above threshold
+    failed_pixels_real = np.count_nonzero(diff_real > 1.0e-5)
+    failed_pixels_imag = np.count_nonzero(diff_imag > 1.0e-5)
+
+    # Compute percentage of pixels in real and imaginary part above threshold
+    percentage_real = failed_pixels_real / tot_pixels_ref
+    percentage_imag = failed_pixels_imag / tot_pixels_ref
+
+    # Check that percentage of pixels above threshold is lower than 0.1 %
+    print('Check that the percentage of pixels in the difference between reference'
+          'and secondary products real parts above the threshold 1.0e-5 is below 0.1 %')
+    assert percentage_real < 0.001
+
+    print('Check that the percentage of pixels in the difference between reference'
+          'and secondary products imaginary parts above the threshold 1.0e-5 is below 0.1 % %')
+    assert percentage_imag < 0.001
```

Comment:
```suggestion
    percent_fail_threshold = 0.001
    for dtype_op, dtype_str in zip([np.real, np.imag], 
                                   ['real', 'imag']):
        diff = (dtype_op(ma_slc_ref) - dtype_op(ma_slc_secsec)) / dype_op(ma_slc_sec)
        percent_fail = diff / tot_pixels_ref
        # Check that percentage of pixels above threshold is lower than 0.1 %
        print(f'Check that the percentage of pixels in the difference between reference'
              'and secondary products {dtype} parts above the threshold {max_diff_threshold} is below {percent_fail_threshold*100} %')
        assert percent_fail < percent_fail_threshold
```
Percentage failed computation moved into loop to avoid unnecessary repeat computation.

Another approach would be run both real and imag tests and saving the error string should either test fail. This is more thorough since both real and imag pixels are checked. Code as follows:
```python
    percent_fail_threshold = 0.001
    any_percent_fail = False
    err_str = []
    for dtype_op, dtype_str in zip([np.real, np.imag], 
                                   ['real', 'imag']):
        diff = (dtype_op(ma_slc_ref) - dtype_op(ma_slc_sec)) / dype_op(ma_slc_sec)
        percent_fail = diff / tot_pixels_ref
        # Check that percentage of pixels above threshold is lower than 0.1 %
        err_str.append(f'Check that the percentage of pixels in the difference between reference'
              'and secondary products {dtype} parts above the threshold {max_diff_threshold} is below {percent_fail_threshold*100} %')
        any_percent_fail = percent_fail < percent_fail_threshold
    assert any_percent_fail, '\n'.join(err_str)
```

Reply:
For the last chunk of code, I am not sure what is the intent with 2 items:

1. `percent_fail` is a matrix but we do want a ballpark number describing the percentage of pixels failing to meet the condition, what is the intent here in your code?
2. In the last line with the `assert` condition, `any_percent_fail` is a matrix again and the `assert` condition will fail. Is this intentional?

Reply:
The current code will only report an error in the imaginary if the real test pass. i.e.:
```
compute real_percentage
assert real_percentage < percentage_threshold

# check imaginary only if previous assert passes
compute imaginary_percentage
assert imaginary_percentage < percentage_threshold
```

The 2nd chunk of code (when properly coded) will test for both real and imaginary. i.e.:
```
for real_imag in [real, imag]:
    if read_imag_percentage > percentage_threshold:
        save that real or imag failed
        no_fail = False
assert no_fail, error_message
```

Apologies for the errors in my code suggestions. The following should address the failing you listed above:
```python
    # declare thresholds
    pixel_diff_threshold = 1e-5
    percent_fail_threshold = 0.001
    
    # container to capture all possible dtype failures
    # tests passes if empty i.e. no failures were captured
    dtype_fails = []

    # iterate to compute and check %age diff in real and imaginary
    for dtype_op, dtype_str in zip([np.real, np.imag], 
                                   ['real', 'imag']):
        # compute diff per pixel
        diff = (dtype_op(ma_slc_ref) - dtype_op(ma_slc_sec)) / dype_op(ma_slc_sec)
        
        # count number that exceed threshold
        n_diff_gt_threshold = np.count_non_zero(diff > pixel_diff_threshold)
        
        # compute % that fail
        percent_fail = n_diff_gt_threshold / tot_pixels_ref
        
        # Check that percentage of pixels above percent_fail_threshold
        if percent_fail > percent_fail_threshold
            dtype_fails.append(dtype_str)

    # check no fails occurred by check of
    err_str = f'Percentage of pixels in the difference between reference and '
               'secondary products parts above the threshold {pixel_diff_threshold} '
               'is above {percent_fail_threshold*100}% for: '
    assert len(dtype_fails) == 0, err_str + ','.join(dtype_fails)
```